classes/DataWarehouse/Visualization/AppKernelChart.php

```diff
@@ -0,0 +1,1075 @@
+                        }
+                        else {
+                            $t0-=12*60*60;
+                        }
+                        if($i1!==count($controlStatus)-1 && count($controlStatus)>1) {
+                            $t1+=0.5*($times[$i1+1]-$times[$i1]);
+                        }
+                        else {
+                            $t1+=12*60*60;
+                        }
+
+                        if($i0!=$i1) {
+                            $this->_chart['layout']['shapes'][] = array(
+                                'name' => 'Out of Control',
+                                'fillcolor' => self::UNDERPERFORMING_COLOR,
+                                'layer' => 'below',
+                                'line' => array(
+                                    'width' => 0
+                                ),
+                                'type' => 'rect',
+                                'xref' => $this->_swapXY ? 'paper' : 'x',
+                                'yref' => $this->_swapXY ? 'y' : 'paper',
+                                'x0' => $this->_swapXY ? 0 : date('Y-m-d H:i:s', $t0),
+                                'x1' => $this->_swapXY ? 1 : date('Y-m-d H:i:s', $t1),
+                                'y0' => $this->_swapXY ? date('Y-m-d H:i:s', $t0) : 0,
+                                'y1' => $this->_swapXY ? date('Y-m-d H:i:s', $t1) : 1,
+                            );
+                        }
+                    }
+                    if($i>=count($controlStatus)) {
+                        break;
+                    }
+
+                    if($controlStatus[$i]==='over_performing') {
+                        $i0=$i;
+                        while($i<count($controlStatus)&&$controlStatus[$i]==='over_performing'){
+                            $i++;
+                        }
+                        $i1=$i;
+                        if($i1>=count($controlStatus)) {
+                            $i1=count($controlStatus)-1;
+                        }
+
+                        $t0=$times[$i0];
+                        $t1=$times[$i1];
+                        if($i0!==0) {
+                            $t0-=0.5*($times[$i0]-$times[$i0-1]);
+                        } else {
+                            $t0-=12*60*60;
+                        }
+                        if($i1!==count($controlStatus)-1 && count($controlStatus)>1) {
+                            $t1+=0.5*($times[$i1+1]-$times[$i1]);
+                        } else {
+                            $t1+=12*60*60;
+                        }
+
+                        if($i0!=$i1) {
+                            $this->_chart['layout']['shapes'][] = array(
+                                'name' => 'Better than Control',
+                                'fillcolor' => self::OVERPERFORMING_COLOR,
+                                'layer' => 'below',
+                                'line' => array(
```

Comment:
Lots of duplicated code in this function...maybe good to refactor in another PR later.